internal/pkg/aws/route53/route53.go

```diff
@@ -20,47 +23,130 @@ const (
 )
 
 type api interface {
-	ListHostedZonesByName(in *route53.ListHostedZonesByNameInput) (*route53.ListHostedZonesByNameOutput, error)
+	ListHostedZonesByName(*route53.ListHostedZonesByNameInput) (*route53.ListHostedZonesByNameOutput, error)
+	ListResourceRecordSets(*route53.ListResourceRecordSetsInput) (*route53.ListResourceRecordSetsOutput, error)
+}
+
+type nameserverResolver interface {
+	LookupNS(ctx context.Context, name string) ([]*net.NS, error)
 }
 
 // Route53 wraps an Route53 client.
 type Route53 struct {
 	client api
+	dns    nameserverResolver
+
+	hostedZoneIDFor map[string]string
 }
 
 // New returns a Route53 struct configured against the input session.
 func New(s *session.Session) *Route53 {
 	return &Route53{
-		client: route53.New(s, aws.NewConfig().WithRegion(route53Region)),
+		client:          route53.New(s, aws.NewConfig().WithRegion(route53Region)),
+		dns:             new(net.Resolver),
+		hostedZoneIDFor: make(map[string]string),
 	}
 }
 
 // DomainHostedZoneID returns the Hosted Zone ID of a domain.
-func (r *Route53) DomainHostedZoneID(domainName string) (string, error) {
+func (r53 *Route53) DomainHostedZoneID(domainName string) (string, error) {
+	if id, ok := r53.hostedZoneIDFor[domainName]; ok {
+		return id, nil
+	}
+
 	in := &route53.ListHostedZonesByNameInput{DNSName: aws.String(domainName)}
-	resp, err := r.client.ListHostedZonesByName(in)
+	resp, err := r53.client.ListHostedZonesByName(in)
 	if err != nil {
 		return "", fmt.Errorf("list hosted zone for %s: %w", domainName, err)
 	}
 	for {
 		hostedZones := filterHostedZones(resp.HostedZones, matchesDomain(domainName))
 		if len(hostedZones) > 0 {
 			// return the first match.
-			return strings.TrimPrefix(aws.StringValue(hostedZones[0].Id), "/hostedzone/"), nil
+			id := strings.TrimPrefix(aws.StringValue(hostedZones[0].Id), "/hostedzone/")
+			r53.hostedZoneIDFor[domainName] = id
+			return id, nil
 		}
 		if !aws.BoolValue(resp.IsTruncated) {
 			return "", &ErrDomainHostedZoneNotFound{
 				domainName: domainName,
 			}
 		}
 		in = &route53.ListHostedZonesByNameInput{DNSName: resp.NextDNSName, HostedZoneId: resp.NextHostedZoneId}
-		resp, err = r.client.ListHostedZonesByName(in)
+		resp, err = r53.client.ListHostedZonesByName(in)
 		if err != nil {
 			return "", fmt.Errorf("list hosted zone for %s: %w", domainName, err)
 		}
 	}
 }
 
+// ValidateDomainOwnership returns nil if the NS records associated with the domain name matches the NS records of the
+// route53 hosted zone for the domain.
+// If there are missing NS records returns ErrUnmatchedNSRecords.
+func (r53 *Route53) ValidateDomainOwnership(domainName string) error {
+	hzID, err := r53.DomainHostedZoneID(domainName)
+	if err != nil {
+		return err
+	}
+
+	wanted, err := r53.listHostedZoneNSRecords(domainName, hzID)
+	if err != nil {
+		return err
+	}
+
+	actual, err := r53.lookupNSRecords(domainName)
+	if err != nil {
+		return err
+	}
+
+	if !eqStringSlices(wanted, actual) {
```

Comment:
Can `wanted` be a subset of `actual` instead of a equal set?

Reply:
Makes sense to me! We want the `actual` to contain only NS records from `wanted`, so that all the NS points to the R53 servers.